### PR TITLE
Manifest Generator

### PR DESCRIPTION
**Pull request by slucero**
### Goals

* Implement prompting and generation of a top-level manifest file capturing important project information to inform further scaffolding
* Implement generation support for Buildkite `pipeline.yml` files within projects
* Implement support for code quality scanning within projects
* Define a migration path for existing projects to upgrade to newly generated standards

#### Manifest Generator 

Execute with:

```
export DEBUG=web-starter:manifest
yo web-starter:manifest
```

* [x] Prompt for top-level project information
* [x] Prompt for project repository information
* [x] Prompt for project environment information
* [x] Prompt for project deployment configurations
* [x] Generate a manifest file documenting captured information

#### Buildkite Pipeline Configuration

Execute with:

```
export DEBUG=web-starter:buildkite-pipeline
yo web-starter:buildkite-pipeline
```

* [x] Generate a project-specific `pipeline.yml` file using project configuration
* [x] Customize deployment steps based on project configuration
    * [x] Generate deployment configuration for Capistrano deployments
    * [x] Generate deployment configuration for Artifact deployments
* [x] Generate required Buildkite scripts and directories
* [x] Generate `docker-compose.buildkite.yml` file

#### Upgrade Path

* [ ] Define and document upgrade paths to add code quality and other improvements into existing projects
* [ ] Document instructions for the creation of new projects

### Notes

* [Example manifest file output](https://gist.github.com/slucero/9aa67438ddfd1dd42042cc9bbd942a29)

#### Generating a new project

```bash
f1 init --next
```
or
```
npx -p yo -p generator-web-starter@next yo web-starter
```

As of right now, the manifest and Buildkite pipeline generators are not incorporated into the project creation generator, so they will need to be executed separately using the commands below.

#### Generating the project manifest file

This generator will run through a series of prompts to gather overall project details and allow for configuration of:

* Git repositories
* Deployment environments
* Deployment workflows

```bash
npx -p yo -p generator-web-starter@next yo web-starter:manifest
```

#### Generating a Buildkite `pipeline.yml` file

Generation of the `pipeline.yml` file and related dependencies may be executed using this generator, but **an existing project manifest file is required** from a previous execution of the manifest generator as instructed above.

```bash
npx -p yo -p generator-web-starter@next yo web-starter:buildkite-pipeline
```

### CODE REVIEW

File: src/manifest/deployment/index.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  },
+];
+
+/**
+ * A Manifest sub-generator responsible for prompting and configuration of deployments.
+ *
+ * This generator is dependent on both the Environment and Repository sub-generators as well.
+ *
+ * @class Deployment
+ * @extends {SubGenerator}
```
Comment (gustavderdrache) [inline src/manifest/deployment/index.ts:30]: Most of these JSDoc annotations aren't needed - TypeScript knows from the `extends` declaration what this is. (This also applies to `@export` and `@memberof`.)

```suggestion
```

File: src/manifest/deployment/index.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+  }
+
+  /**
+   * Create a standardized prompt for selecting a configured environment.
+   *
+   * @param {Partial<ListQuestion>} options
+   *   Override options to contextualize the presentation of the prompt.
+   * @returns {ListQuestion}
+   *   The contextualized prompt ready to be used.
+   * @memberof Deployment
```
Comment (gustavderdrache) [inline src/manifest/deployment/index.ts:109]: TypeScript ignores the type annotations in these comments - the method declaration is the source of truth here.

```suggestion
   * @param options Override options to contextualize the presentation of the prompt.
   * @returns The contextualized prompt ready to be used.
```

**Comment by slucero** [comment]
Thanks for the feedback @gustavderdrache! The class-level use of generics is an intriguing application I hadn't come to realize yet, but I like it. And it's a nice consolidation of a lot the boilerplate I wanted to rip out of there.

Many of the other items look like they may potentially apply more widely than the individual instances that were commented on, so I'll handle those in separate commits rather than the suggestions on here directly.

File: src/manifest/index.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  /**
+   * Execute the writing phase of this generator.
+   *
+   * @memberof Manifest
+   */
+  writing() {
+    // Write all manifest content to a YAML file.
+    this.fs.write('.f1-manifest.yml', YAML.stringify(this.manifest));
```
[the remark this answers is not in the document]
Comment (slucero) [reply src/manifest/index.ts:166]: It does write it in the correct place in all of my testing, but you're right that it should likely be more explicit. I may run into an issue with it if I weren't running the generator from within the top-level directory.

File: src/manifest/deployment/index.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+ *
+ * This generator is dependent on both the Environment and Repository sub-generators as well.
+ */
+class Deployment extends SubGenerator<DeploymentDefinition> {
+  private repositories: RepositoryCollection = {};
+  private environments: EnvironmentCollection = {};
+  protected configName = 'deployments';
+  protected editPrompt = 'Would you like to update your deployments?';
+
+  _setResources(resources: Record<string, ResourceCollection>): void {
```
Comment (gustavderdrache) [inline src/manifest/deployment/index.ts:31]: Now that we've addressed the types of the subgenerators, we can fix this method. This method is only used by this generator, so we can remove the signature from the parent class (and its implementations in `Repository` and `Environment`), and either split this method into two (`_setRepositories` and `_setEnvironments`) or update the signature to be `_setResources(repositories, environments)`.

**Comment by slucero** [comment]
I've published a test build of this as pre-release `v2.6.0-0` on branch `release-next`.

This pre-release may be tested for general project generation using either of these commands:
```bash
f1 init --next
```
or
```
npx -p yo -p generator-web-starter@next yo web-starter
```

The additional generators may be tested using the following commands in order and completing the prompts appropriately:
```bash
npx -p yo -p generator-web-starter@next yo web-starter:manifest
```
```bash
npx -p yo -p generator-web-starter@next yo web-starter:buildkite-pipeline
```

**Comment by slucero** [comment]
### Testing Notes

While trying to test and configure a new site for Pantheon using the new prerelease, I identified the following manual changes needed for clean-up after running all three commands:

* The `pipeline.yml` file included deployment steps for both Capistrano and an Artifact deployment even though no configuration was selected for a Capistrano deployment.
• The `pipeline.yml` file includes the step for running Code Climate which will need to be manually removed. This should be excluded for now with the final code quality solution incorporated as a follow-up from work in this PR.

**Pull request merged**